airbyte-protocol/models/src/main/java/io/airbyte/protocol/models/JsonSchemaType.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (c) 2021 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.protocol.models;
+
+import java.util.HashMap;
+
+public class JsonSchemaType {
+
+  public static final String TYPE = "type";
+  public static final String FORMAT = "format";
+  public static final String DATE_TIME = "date-time";
+  public static final String DATE = "date";
+  public static final String TIME = "time";
+  public static final String TIMESTAMP_WITH_TIMEZONE = "timestamp_with_timezone";
+  public static final String TIMESTAMP_WITHOUT_TIMEZONE = "timestamp";
+  public static final String CONTENT_ENCODING = "contentEncoding";
+  public static final String BASE_64 = "base64";
+  public static final String AIRBYTE_TYPE = "airbyte_type";
+
+  public static final JsonSchemaType STRING = new JsonSchemaType.Builder().withType(JsonSchemaPrimitive.STRING).build();
+  public static final JsonSchemaType NUMBER = new JsonSchemaType.Builder().withType(JsonSchemaPrimitive.NUMBER).build();
+  public static final JsonSchemaType BOOLEAN = new JsonSchemaType.Builder().withType(JsonSchemaPrimitive.BOOLEAN).build();
+  public static final JsonSchemaType OBJECT = new JsonSchemaType.Builder().withType(JsonSchemaPrimitive.OBJECT).build();
+  public static final JsonSchemaType ARRAY = new JsonSchemaType.Builder().withType(JsonSchemaPrimitive.ARRAY).build();
+  public static final JsonSchemaType NULL = new JsonSchemaType.Builder().withType(JsonSchemaPrimitive.NULL).build();
+  public static final JsonSchemaType STRING_BASE_64 =
+      new JsonSchemaType.Builder().withType(JsonSchemaPrimitive.STRING).withContentEncoding(BASE_64).build();
+
+  private final HashMap<String, String> jsonSchemaTypeMap = new HashMap<>();
+
+  public HashMap<String, String> getJsonSchemaTypeMap() {
+    return jsonSchemaTypeMap;
+  }
+
+  public static class Builder {
+
+    private final JsonSchemaType jsonSchemaType;
+
+    public Builder() {
+      jsonSchemaType = new JsonSchemaType();
+    }
+
+    public Builder withType(JsonSchemaPrimitive primitive) {
+      jsonSchemaType.jsonSchemaTypeMap.put(TYPE, primitive.name().toLowerCase());
+      return this;
+    }
+
+    public Builder withFormat(String value) {
+      jsonSchemaType.jsonSchemaTypeMap.put(FORMAT, value);
+      return this;
+    }
+
+    public Builder withContentEncoding(String value) {
+      jsonSchemaType.jsonSchemaTypeMap.put(CONTENT_ENCODING, value);
+      return this;
+    }
+
+    public Builder withAirbyteType(String value) {
+      jsonSchemaType.jsonSchemaTypeMap.put(AIRBYTE_TYPE, value);
+      return this;
+    }
+
+    public JsonSchemaType build() {
+      return jsonSchemaType;
+    }
+
+  }
```

Comment:
In the builder pattern, the object is usually only constructed at the end. We don't want to manipulate the object as we build it. The `TYPE` parameter is required, right? Currently the builder may create an invalid type object if the client do this: `JsonSchemaType type = new JsonSchemaType.Builder().build()`. So the constructor of the builder should take in a type parameter. Also the internal map of in `JsonSchemaType` should be immutable.

I think a better way to implement this class and the build is that:
1. The builder has a reference to a map builder, instead of a `JsonSchemaType`.
2. Each of the `with-` method adds new entries to the map builder, instead of updating `jsonSchemaType.jsonSchemaTypeMap` directly.
3. Add a constructor to the `JsonSchemaType` that takes in a `Map` parameter to initialize the `jsonSchemaTypeMap` value. This constructor should be `private` and only accessible through the builder. We don't want client to pass in maps with invalid values.
4. The `build` method in the builder creates the map from the map builder, and passes it to the constructor of the `JsonSchemaType` class to create an instance.
5. You can even mark the constructor of the builder as private, and add a static `builder` method to the `JsonSchemaType` class so that the only way to get a builder is by `JsonSchemaType.builder()`, which  may be more straightforward because people may not know there is a builder class inside the type class.

To summarize all the above, the type class will look like this:

```java
/*
 * Copyright (c) 2021 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.protocol.models;

import com.google.common.collect.ImmutableMap;
import java.util.Map;

public class JsonSchemaType {

  public static final String TYPE = "type";
  public static final String FORMAT = "format";
  public static final String DATE_TIME = "date-time";
  public static final String DATE = "date";
  public static final String TIME = "time";
  public static final String TIMESTAMP_WITH_TIMEZONE = "timestamp_with_timezone";
  public static final String TIMESTAMP_WITHOUT_TIMEZONE = "timestamp";
  public static final String CONTENT_ENCODING = "contentEncoding";
  public static final String BASE_64 = "base64";
  public static final String AIRBYTE_TYPE = "airbyte_type";

  public static final JsonSchemaType STRING = JsonSchemaType.builder(JsonSchemaPrimitive.STRING).build();
  public static final JsonSchemaType NUMBER = JsonSchemaType.builder(JsonSchemaPrimitive.NUMBER).build();
  public static final JsonSchemaType BOOLEAN = JsonSchemaType.builder(JsonSchemaPrimitive.BOOLEAN).build();
  public static final JsonSchemaType OBJECT = JsonSchemaType.builder(JsonSchemaPrimitive.OBJECT).build();
  public static final JsonSchemaType ARRAY = JsonSchemaType.builder(JsonSchemaPrimitive.ARRAY).build();
  public static final JsonSchemaType NULL = JsonSchemaType.builder(JsonSchemaPrimitive.NULL).build();
  public static final JsonSchemaType STRING_BASE_64 = JsonSchemaType.builder(JsonSchemaPrimitive.STRING).withContentEncoding(BASE_64).build();

  private final Map<String, String> jsonSchemaTypeMap;

  private JsonSchemaType(final Map<String, String> jsonSchemaTypeMap) {
    this.jsonSchemaTypeMap = jsonSchemaTypeMap;
  }

  public static Builder builder(final JsonSchemaPrimitive type) {
    return new Builder(type);
  }

  public Map<String, String> getJsonSchemaTypeMap() {
    return jsonSchemaTypeMap;
  }

  public static class Builder {

    private final ImmutableMap.Builder<String, String> typeMapBuilder;

    private Builder(final JsonSchemaPrimitive type) {
      typeMapBuilder = ImmutableMap.builder();
      typeMapBuilder.put(TYPE, type.name().toLowerCase());
    }

    public Builder withFormat(String value) {
      typeMapBuilder.put(FORMAT, value);
      return this;
    }

    public Builder withContentEncoding(String value) {
      typeMapBuilder.put(CONTENT_ENCODING, value);
      return this;
    }

    public Builder withAirbyteType(String value) {
      typeMapBuilder.put(AIRBYTE_TYPE, value);
      return this;
    }

    public JsonSchemaType build() {
      return new JsonSchemaType(typeMapBuilder.build());
    }

  }

}
```

Reply:
@tuliren great suggestion. Updated JsonSchemaType